src/BuildLoop.cpp

```diff
@@ -150,27 +160,237 @@ static void SignalMainThreadToStartCleaningUp(BuildQueue *queue)
     MutexUnlock(&queue->m_BuildFinishedMutex);
 }
 
-static void AdvanceNode(BuildQueue *queue, ThreadState *thread_state, RuntimeNode *node, Mutex *queue_lock)
+static void FinishNode(BuildQueue* queue, RuntimeNode* node)
+{
+    CheckHasLock(&queue->m_Lock);
+
+    node->m_Finished = true;
+    queue->m_FinishedNodeCount++;
+
+    if (RuntimeNodeIsExplicitlyRequested(node))
+        queue->m_FinishedRequestedNodeCount++;
+
+    RuntimeNodeFlagInactive(node);
+    if (queue->m_FinishedRequestedNodeCount == queue->m_Config.m_AmountOfRuntimeNodesSpecificallyRequested)
+        SignalMainThreadToStartCleaningUp(queue);
+
+    EnqueueDependeesWhoMightNowHaveBecomeReadyToRun(queue, node);
+}
+
+static bool IsNodeCacheableByLeafInputsAndCachingEnabled(BuildQueue* queue, RuntimeNode* node)
+{
+    if (!queue->m_Config.m_AttemptCacheReads && !queue->m_Config.m_AttemptCacheWrites)
+        return false;
+    return 0 != (node->m_DagNode->m_Flags & Frozen::DagNode::kFlagCacheableByLeafInputs);
+}
+
+static void AttemptCacheWrite(BuildQueue* queue, ThreadState* thread_state, RuntimeNode* node)
+{
+    CheckDoesNotHaveLock(&queue->m_Lock);
+
+    uint64_t time_exec_started = TimerGet();
+
+    char path[kMaxPathLength];
+    char digestString[kDigestStringSize];
+    DigestToString(digestString, node->m_CurrentLeafInputSignature->digest);
+    snprintf(path, sizeof(path), "%s", digestString);
+    FILE *sig = fopen(path, "w");
```

Comment:
Isn't digestString already a `const char *`? Could we just fopen that without `path`?

Reply:
Indeed. This should be simpler. An earlier version of this code put the path into a directory, so assembling a path made sense, but it no longer does. I'll fix.